Bonus guide: Mempool

VajraOfIndra opened the pull request:
#### What

This is a proposal to add a bonus guide to install Mempool.

### Why

It is a popular Bitcoin blockchain explorer and mempool visualizer and can be used as an alternative or complement to the BTC RPC Explorer which is installed in the main guide.

Fixes #840 and #651 

#### How

- Bonus guide at /bonus/+bitcoin/mempool.md
- Updated /bonus/+bitcoin/index.md to add a reference to the Mempool guide (under 'Privacy', would another heading be more suitable?)

- The guide requires in-depth review, in particular the nginx configuration section. See Issue #840 for a list of on-going questions and things to review/sort out.

#### Scope

- [ ] significant change to core configuration
- [x] independent bonus guide
- [ ] simple bug fix

Fixes #840 and #651 

#### Test & maintenance

Test: 
* Follow the guide and run Mempool in your browser
* Check nginx config, is it secure?

Maintenance: Checking breaking changes at each Mempool update. Update of the 'Upgrade' section.

Maintenance:

#### Animated GIF (optional)

![mempool](https://media.giphy.com/media/XSVyi8jXJcv4ZEvOas/giphy-downsized.gif)

RooSoft commented on bonus/bitcoin/mempool.md:
```diff
@@ -0,0 +1,623 @@
+      "NETWORK": "mainnet",
+      "BACKEND": "electrum",
+      "HTTP_PORT": 8999,
+      "NETWORK": "mainnet",
```
The next 3 lines are duplicates of the first 3

RooSoft commented on bonus/bitcoin/mempool.md:
````diff
@@ -0,0 +1,623 @@
+* Copy the conf file dedicated to the Mempool website in the nginx `snippets` directory
+
+  ```sh
+  $ sudo rsync -av /home/mempool/mempool/nginx-mempool.conf /etc/nginx/snippets
````
Pretty sure only `nginx-mempool.conf` is required here....

Removed both `fastcgi-php.conf` and `snakeoil.conf` and it is working just fine right now.

VajraOfIndra replied:
These 2 files are installed when nginx is installed.. do you know what they;re used for?

Better deleteing them or changed their names to .bak?

VajraOfIndra replied:
Ok

RooSoft reviewed (changes requested):
Made some minor recommendations

VajraOfIndra commented on bonus/bitcoin/mempool.md:
````diff
@@ -0,0 +1,623 @@
+  > v16.13.1
+  ```
+
+* If NodeJS is not installed, follow [this guide](https://raspibolt.org/btcrpcexplorer.html#install-nodejs) to install it. If the version is v14 or older, update it following [this tutorial](https://phoenixnap.com/kb/update-node-js-version){:target="_blank"}.
````
- Node.js version: yes, I agree we should make a PR to use v16 in the Explorer guide

- Spelling: great point, I'll change it to the right spelling in this guide. We should make another PR to change the spelling in the other guide.

VajraOfIndra commented on bonus/bitcoin/mempool.md:
````diff
@@ -0,0 +1,623 @@
+* Enter the following commands in the shell and exit. The instructions to enter in the MariaDB shell with start with "MDB$"
+
+  ```sql
+  MDB$ drop database mempool;
````
Should we remove that line instead? if not, could you suggest what lines we should add to explain how to figure it out?

RooSoft replied:
Agree it could be completely removed to simplify the document

VajraOfIndra replied:
Ok, it's removed now

VajraOfIndra commented on bonus/bitcoin/mempool.md:
```diff
@@ -0,0 +1,623 @@
+    "MEMPOOL": {
+      "NETWORK": "mainnet",
+      "BACKEND": "electrum",
+      "HTTP_PORT": 8999,
```
Well spotted, fixed!

VajraOfIndra commented on bonus/bitcoin/mempool.md:
````diff
@@ -0,0 +1,623 @@
+    }
+  }
+  ```
+
````
Great idea!
I couldn't use chown with the "mempool" user (is that normal..?) so I placed the command once we had exited the mempool user (see new heading just after the frontend section).

Happy to change it to an earlier position if there is a way to do it from the mempool user directly

VajraOfIndra replied:
Did it work on yours? It didn't work on mine.. any idea what i could ahve done wrong?

RooSoft replied:
Can you send the result of `ls -lsa` in that folder?

VajraOfIndra commented on bonus/bitcoin/mempool.md:
````diff
@@ -0,0 +1,623 @@
+  ```sh
+  $ cd frontend
+  $ npm install --prod
+  $ npm run build
````
does v2.2 resolves to v2.2.0? and is it because the latest version v2.2.2 leads to some errors, right?

I suppose the git checkout could be done just after cloning the repo as is recommended in their repo? (except that they recommend to checkout the latest release and will use v2.2 instead.

See my changes at lines 132-135, let me know if that works well or not.

VajraOfIndra replied:
So what's the relationship between releases (v2.2.0, v2.2.1, v2.2.2) and branches (v2.2) then? I'm really confused now :)

VajraOfIndra replied:
> Looking pretty good!



> Looking pretty good!

All good now I think, PTAL (and resolve the comments if you're satisfied with the changes). I just need to do some reading on how git releases and branches and tags and all that good stuff work as I'm quite confused by it right now :D

RooSoft replied:
Releases merely are tags on a specific commit along with downloadable binaries of the compiled app at that point.

VajraOfIndra commented on bonus/bitcoin/mempool.md:
````diff
@@ -0,0 +1,623 @@
+* Replace the main nginx configuration file
+
+  ```sh
+  $ sudo mv /etc/nginx/nginx.conf /etc/nginx/nginx.conf.bak2
````
We back up the config file created by the RaspiBolt main guide. The guide laready led us to create a nginx.conf.bak (which is a backup of the default conf file that comes with nginx). Hence the idea of backing this one up with the name .bak2. We would end with:
* `nginx.conf `= the conf file required for the RaspiBolt reverse proxy servers + Mempool web server
* `nginx.conf.bak `= the default nginx conf file that comes when installing nginx
* `nginx.conf.bak2` = a backup of the conf file created when installing theraspibolt for the reverse proxy servers

But we could replace `nginx.conf.bak2` by `raspibolt-nginx.conf.bak' for example to have a more semantically menaningful name, wdyt? (I used this new name for now)

VajraOfIndra replied:
I see. I think the new name I gave it helps. It reminds people that we're bakcing up the nginx config that is created for the RaspiBolt setup (i.e. for the reverse proxy servers).

VajraOfIndra commented:
> Made some minor recommendations

Thanks for the recommendations! I ammended accordingly, PTAL. Also, see replies above, I have a couple of questions.

RooSoft reviewed (approved):
Looking pretty good!

VajraOfIndra commented:
@Stadicus: the Mempool guide is now ready for final review after a very helpful review and improvement suggestions by @RooSoft.

The whole guide needs to be reviewed and will benefit from your suggestions, but I think the nginx config section needs particular/extra attention as I'm not entirely sure it's optimally configured as it is now.

Stadicus commented:
Great work. I'll try to finish the SCB/GitHub backup pr first and then will give this guide a go.

cyberunknown commented:
This guide works well except this:
> /etc/nginx/snippets/nginx-mempool.conf
`allow 192.168.0.0/16;`
`deny all;`

Having problems on language change to EN (Forbidden 403) and logo of mempool not loading.
In official mempool installation guide this not mentioned. So I think best option is delete this from guide.

VajraOfIndra commented:
> This guide works well except this:
> 
> > /etc/nginx/snippets/nginx-mempool.conf
> > `allow 192.168.0.0/16;`
> > `deny all;`
> 
> Having problems on language change to EN (Forbidden 403) and logo of mempool not loading. In official mempool installation guide this not mentioned. So I think best option is delete this from guide.

Well spotted, thanks! that was something I added as a test and forgot to remove later when I realised it was not needed. I've now removed that entire section. PTAL.

> Also tested this guide with mempool v2.3.1. All works well.

Great to know. I've updated the guide to v2.3.1(basically git checkout v2.3.1 instead of git checkout v2.2) but didn't test it.

RooSoft reviewed (approved):
All good to me after reviewing changes

VajraOfIndra commented:
Let4s go to master! :)
